llvm/test/CodeGen/RISCV/rvv/fixed-vectors-binop-splats.ll

```diff
@@ -0,0 +1,592 @@
+; RV32-NEXT:    addi a0, sp, 8
+; RV32-NEXT:    vsetivli zero, 2, e64, m1, ta, ma
+; RV32-NEXT:    vlse64.v v8, (a0), zero
+; RV32-NEXT:    sw a3, 4(sp)
+; RV32-NEXT:    sw a2, 0(sp)
+; RV32-NEXT:    mv a0, sp
+; RV32-NEXT:    vlse64.v v9, (a0), zero
+; RV32-NEXT:    vadd.vv v9, v8, v9
+; RV32-NEXT:    vrgather.vi v8, v9, 0
+; RV32-NEXT:    addi sp, sp, 16
+; RV32-NEXT:    ret
+;
+; RV64-LABEL: v2i64:
+; RV64:       # %bb.0:
+; RV64-NEXT:    vsetivli zero, 2, e64, m1, ta, ma
+; RV64-NEXT:    vmv.v.x v8, a0
+; RV64-NEXT:    vadd.vx v9, v8, a1
+; RV64-NEXT:    vrgather.vi v8, v9, 0
+; RV64-NEXT:    ret
+  %head.x = insertelement <2 x i64> poison, i64 %x, i32 0
+  %splat.x = shufflevector <2 x i64> %head.x, <2 x i64> poison, <2 x i32> zeroinitializer
+  %head.y = insertelement <2 x i64> poison, i64 %y, i32 0
+  %splat.y = shufflevector <2 x i64> %head.y, <2 x i64> poison, <2 x i32> zeroinitializer
+  %v = add <2 x i64> %splat.x, %splat.y
+  ret <2 x i64> %v
+}
+
+define <4 x i64> @v4i64(i64 %x, i64 %y) {
+; RV32-LABEL: v4i64:
+; RV32:       # %bb.0:
+; RV32-NEXT:    addi sp, sp, -16
+; RV32-NEXT:    .cfi_def_cfa_offset 16
+; RV32-NEXT:    sw a1, 12(sp)
+; RV32-NEXT:    sw a0, 8(sp)
+; RV32-NEXT:    addi a0, sp, 8
+; RV32-NEXT:    vsetivli zero, 4, e64, m2, ta, ma
+; RV32-NEXT:    vlse64.v v8, (a0), zero
+; RV32-NEXT:    sw a3, 4(sp)
+; RV32-NEXT:    sw a2, 0(sp)
+; RV32-NEXT:    mv a0, sp
+; RV32-NEXT:    vlse64.v v10, (a0), zero
+; RV32-NEXT:    vadd.vv v10, v8, v10
+; RV32-NEXT:    vrgather.vi v8, v10, 0
+; RV32-NEXT:    addi sp, sp, 16
+; RV32-NEXT:    ret
+;
+; RV64-LABEL: v4i64:
+; RV64:       # %bb.0:
+; RV64-NEXT:    vsetivli zero, 4, e64, m2, ta, ma
+; RV64-NEXT:    vmv.v.x v8, a0
+; RV64-NEXT:    vadd.vx v10, v8, a1
+; RV64-NEXT:    vrgather.vi v8, v10, 0
+; RV64-NEXT:    ret
+  %head.x = insertelement <4 x i64> poison, i64 %x, i32 0
+  %splat.x = shufflevector <4 x i64> %head.x, <4 x i64> poison, <4 x i32> zeroinitializer
+  %head.y = insertelement <4 x i64> poison, i64 %y, i32 0
+  %splat.y = shufflevector <4 x i64> %head.y, <4 x i64> poison, <4 x i32> zeroinitializer
+  %v = add <4 x i64> %splat.x, %splat.y
+  ret <4 x i64> %v
+}
+
+define <8 x i64> @v8i64(i64 %x, i64 %y) {
+; RV32-LABEL: v8i64:
+; RV32:       # %bb.0:
+; RV32-NEXT:    addi sp, sp, -16
+; RV32-NEXT:    .cfi_def_cfa_offset 16
+; RV32-NEXT:    sw a1, 12(sp)
+; RV32-NEXT:    sw a0, 8(sp)
+; RV32-NEXT:    addi a0, sp, 8
+; RV32-NEXT:    vsetivli zero, 8, e64, m4, ta, ma
+; RV32-NEXT:    vlse64.v v8, (a0), zero
+; RV32-NEXT:    sw a3, 4(sp)
+; RV32-NEXT:    sw a2, 0(sp)
+; RV32-NEXT:    mv a0, sp
+; RV32-NEXT:    vlse64.v v12, (a0), zero
+; RV32-NEXT:    vadd.vv v12, v8, v12
+; RV32-NEXT:    vrgather.vi v8, v12, 0
+; RV32-NEXT:    addi sp, sp, 16
+; RV32-NEXT:    ret
+;
+; RV64-LABEL: v8i64:
+; RV64:       # %bb.0:
+; RV64-NEXT:    vsetivli zero, 8, e64, m4, ta, ma
+; RV64-NEXT:    vmv.v.x v8, a0
+; RV64-NEXT:    vadd.vx v12, v8, a1
+; RV64-NEXT:    vrgather.vi v8, v12, 0
```

Comment:
Not sure what's going on here: something's recognising that it's a splat, but it's splatting the result after the vadd?

Reply:
If I'm reading this right, its trying to move the result of the add `v12` into `v8`. It's doing this move using vrgather. I think this could have done `vadd.vx v8, v8, a1` and removed the need for the vrgather?

Reply:
Yeah it should just be `vadd.vx v8, v8, a1`, but the weird thing is that the `vrgather.vi` is actually splatting the first element of `v12` into `v8`. Which is weird because the the input IR doesn't say anything about splatting `%v`.

So it must somehow know that the result of `vadd.vx v12, v8, a1` is a splatted vector, and then it splats the first element again anyway which is a no-op?

Reply:
> So it must somehow know that the result of vadd.vx v12, v8, a1 is a splatted vector, and then it splats the first element again anyway which is a no-op?

It seems that way.